<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav2)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="50"/>

  <arg name="device" default="$(optenv BLUEFOX_UV_LEFT)"/>
  <arg name="left" default="$(arg device)"/>
  <arg name="right" default="$(optenv BLUEFOX_UV_RIGHT)"/>
  <arg name="left_camera_name" default="mv_$(arg left)"/>
  <arg name="right_camera_name" default="mv_$(arg right)"/>

  <arg name="rate" default="100"/>
  <arg name="camera_name" default="bluefox"/>
  <arg name="camera" default="$(arg camera_name)"/>
  <arg name="frame_id" default="$(arg camera)"/>
  <arg name="calib_url" default="file://$(find mrs_general)/camera_info/bluefox_$(arg uav_name).yaml"/>
  <arg name="fps" default="$(arg rate)"/>
  <arg name="idpf" default="0"/>
  <arg name="aec" default="false"/>
  <arg name="expose_us" default="50"/>
  <arg name="agc" default="false"/>
  <arg name="gain_db" default="0.0"/>
  <arg name="cbm" default="0"/>
  <arg name="ctm" default="1"/>
  <arg name="dcfm" default="0"/>
  <arg name="hdr" default="false"/>
  <arg name="wbp" default="-1"/>
  <arg name="request" default="0"/>
  <arg name="mm" default="0"/>
  <arg name="jpeg_quality" default="90"/>


  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="true"/>
  <!-- <arg name="gui" default="false"/> -->

  <arg name="trajectoryControl" default="true"/>

  <arg name="followDistance" default="5"/>
  <arg name="trajCoeff" default="0.8"/>
  <arg name="yawCoeff" default="0.1"/>
  <arg name="tailingCoeff" default="1.0"/>

  <arg name="publish" default="true"/>
  <arg name="useOdom" default="false"/>
  <arg name="cemeraRotated" default="false"/>
  <arg name="FromVideo" default="false"/>
  <arg name="VideoNumber" default="9"/>
  <arg name="cellSize" default="64"/>
  <arg name="cellOverlay" default="16"/>
  <arg name="camNum" default="0"/>
  <arg name="Delay" default="100"/>
  <arg name="TimeScale" default="1.0"/>

  <arg name="frameMRSid" default="$(optenv MRS_ID)"/>

  <group ns="$(arg uav_name)">
  
    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch"> -->
    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" >
      <param name="num_worker_threads" value="8" />
    </node>

    <node pkg="nodelet" type="nodelet" name="$(arg left_camera_name)" args="$(arg nodelet) bluefox2/SingleNodelet $(arg nodelet_manager)" respawn="true" output="screen" launch-prefix="bash -c 'sleep 5; $0 $@'">
      <param name="camera_name" type="string" value="$(arg left_camera_name)"/>
      <param name="identifier" type="string" value="$(arg left)"/>

      <param name="right/camera_name" type="string" value="$(arg right_camera_name)"/>
      <param name="right/identifier" type="string" value="$(arg right)"/>

      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="fps" type="double" value="$(arg fps)"/>
      <param name="idpf" type="int" value="$(arg idpf)"/>
      <param name="aec" type="bool" value="$(arg aec)"/>
      <param name="expose_us" type="int" value="$(arg expose_us)"/>
      <param name="agc" type="bool" value="$(arg agc)"/>
      <param name="gain_db" type="double" value="$(arg gain_db)"/>
      <param name="cbm" type="int" value="$(arg cbm)"/>
      <param name="ctm" type="int" value="$(arg ctm)"/>
      <param name="dcfm" type="int" value="$(arg dcfm)"/>
      <param name="hdr" type="bool" value="$(arg hdr)"/>
      <param name="wbp" type="int" value="$(arg wbp)"/>
      <param name="request" type="int" value="$(arg request)"/>
      <param name="mm" type="int" value="$(arg mm)"/>

      <param name="image_raw/compressed/jpeg_quality" type="int" value="$(arg jpeg_quality)"/>
      <param name="image_raw/theora/keyframe_frequency" value="60" />
      <param name="image_raw/theora/target_bitrate" value="50000" />
      <param name="image_raw/theora/quality" value="8" />
      <param name="image_raw/theora/optimize_for" value="0" />
    </node>

    <node pkg="rosbag" type="record" name="rosbag_recorder" args="-a --output-name=$(optenv HOME)/bag_files/uvdar_calibration/$(arg device).bag" output="screen"/>

  </group>


</launch>
